### PR TITLE
Exclude types originating in literals from recursion depth limiter check

### LINKED ISSUE #34619: Not reporting error of deep object literal as expected
**TypeScript Version:**  3.6.3 / 3.7.0-dev.20191021

**Search Terms:** type check, no error, large object

**Code**

```ts
interface A {
  b: B[];
}

interface B {
  c: C;
}

interface C {
  d: D[];
}

interface D {
  e: E[];
}

interface E {
  f: F[];
}

interface F {
  g: G;
}

interface G {
  h: H[];
}

interface H {
  i: string;
}

const x: A = {
  b: [
    {
      c: {
        d: [
          {
            e: [
              {
                f: [
                  {
                    g: {
                      h: [
                        {
                          // intellisense still works, but type check fails here
                          // i: '',
                        },
                      ],
                    },
                  },
                ],
              },
            ],
          },
        ],
      },
    },
  ],
};
```

**Expected behavior:**

Reporting error as property `i` is missing.

**Actual behavior:**

No error.

**Playground Link:** http://www.typescriptlang.org/play/?ts=Nightly&ssl=1&ssc=1&pln=61&pc=1#code/JYOwLgpgTgZghgYwgAgILIN4ChnIEYBcyAQgNoC6A3FgL5ZaiSyIrGY7IJEDC1dD4aPCTJu7XABMiAEQp96jIS2TTxyCEQCic2gsHMRmtTCIAxHf0UGUptQHMiAcXkCmwlI7UALIgAkLem7KvmrARADOYFCgdi4IAPYgkcgAHkToALxqhMikHLjYuEWcRIXFRVK5+eUF1TW4GlX19WXNNSZNbc2tXfUOar1tPp2DXT2jbQD0k8hgAJ4ADigIXhAIANbI8MAANuHIq1AQdRP108hhyADkVwA0J6dFNPePNeQvr7jPD73fj+8-Gp-XoArrAt4fYrg5DAgE0ahAA

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #34619.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16272,13 +16272,13 @@ namespace ts {
...
             // We track all object types that have an associated symbol (representing the origin of the type)
-            if (depth >= 5 && type.flags & TypeFlags.Object) {
+            if (depth >= 5 && type.flags & TypeFlags.Object && !isObjectOrArrayLiteralType(type)) {
                 const symbol = type.symbol;
                 if (symbol) {
                     let count = 0;
                     for (let i = 0; i < depth; i++) {
                         const t = stack[i];
-                        if (t.flags & TypeFlags.Object && t.symbol === symbol) {
+                        if (t.flags & TypeFlags.Object && t.symbol === symbol && !isObjectOrArrayLiteralType(t)) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: If two types have the same symbol and one of them is already known not to be an array or object literal type, how could the other possibly be? Under what scenarios does this need to be checked here, too?

Comment (ahejlsberg) [reply src/compiler/checker.ts]: Yeah, you're right.

**Comment by ahejlsberg** [comment]
@typescript-bot run dt
@typescript test this

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
RWC tests are clean. DT tests have one new error in tests for `mergerino` package (7 level deep array literal is now being checked where it wasn't before).

**Comment by weswigham** [comment]
Is there a case where there's an object literal type that's provided as an argument to a symbolless generic (indexed access?) thats somehow nonterminating now?

**Comment by ahejlsberg** [comment]
Don't think so. But if there is, then there is now as well since literals by themselves can't be circular.

**Pull request merged**